### PR TITLE
[WIP] refactor BMM+BMM 

### PR DESCRIPTION
**Pull request by asroy**

### CODE REVIEW

**Review by rosenrodt — COMMENTED** [review]
A few inline comments above. Aside from that - how much does the change both inside and outside of device op impact others' work currently depending on it? cc @ltqin

**Comment by rosenrodt** [comment]
Could you also briefly summarize what changes are done in the description? BMM+Softmax+BMM will follow suit.

File: example/31_batched_gemm_gemm/batched_gemm_gemm_xdl_fp16.cpp [hunk 1]
```diff
@@ -36,43 +35,51 @@ using Col = ck::tensor_layout::gemm::ColumnMajor;
...
+using A0DataType        = F16;
+using B0DataType        = F16;
+using Acc0DataType      = F32;
+using B1DataType        = F16;
+using Acc1DataType      = F32;
+using C1ShuffleDataType = F32;
+using C1DataType        = F16;
 
-using ALayout  = Row;
+using A0Layout = Row;
```
[the remark this answers is not in the document]
Comment (asroy) [reply example/31_batched_gemm_gemm/batched_gemm_gemm_xdl_fp16.cpp:46]: done

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_gemm_xdl_cshuffle.hpp [hunk 2]
```diff
@@ -74,185 +74,176 @@ __global__ void
...
+          typename C1DataType,
+          typename A0ElementwiseOperation,
+          typename B0ElementwiseOperation,
+          typename C0ElementwiseOperation,
           typename B1ElementwiseOperation,
-          typename CElementwiseOperation,
-          GemmSpecialization GemmSpec,
-          index_t NumGemmKPrefetchStage,
+          typename C1ElementwiseOperation,
+          bool Gemm0PadM,
```
[the remark this answers is not in the document]
Comment (asroy) [reply include/ck/tensor_operation/gpu/device/device_batched_gemm_gemm_xdl_cshuffle.hpp]: done, but I alias ```PadM``` to ```Gemm0PadM```, etc, inside the DeviceOp

File: include/ck/tensor_operation/gpu/device/matrix_padder.hpp [hunk 3]
```diff
@@ -198,6 +198,166 @@ struct MatrixPadder : public GemmPadder<GemmSpec, MPerTileType, NPerTileType, KP
...
+    static constexpr auto I0 = Number<0>{};
+    static constexpr auto I1 = Number<1>{};
+    static constexpr auto I2 = Number<2>{};
+    static constexpr auto I3 = Number<3>{};
+
+    template <typename ADesc_MRaw_KRaw>
+    __host__ __device__ constexpr auto
+    PadADescriptor_M_K(const ADesc_MRaw_KRaw& a_desc_mraw_kraw) const
+    {
+        const auto MRaw = a_desc_mraw_kraw.GetLength(I0);
```
[the remark this answers is not in the document]
Comment (asroy) [reply include/ck/tensor_operation/gpu/device/matrix_padder.hpp]: Done. I refacrtored ```PadTensorDescriptor``` to take any dimensional tensor

**Comment by rosenrodt** [comment]
The CI crashes at build stage with the following error log. I restarted the pipeline to see if it's the one-off thing.
```
[2022-08-24T18:41:39.677Z] fatal error: error in backend: Error while trying to spill VGPR0 from class VGPR_32: Cannot scavenge register without an emergency spill slot!
[2022-08-24T18:41:39.677Z] clang-13: error: clang frontend command failed with exit code 70 (use -v to see invocation)
[2022-08-24T18:41:39.677Z] clang version 13.0.0 (https://github.com/RadeonOpenCompute/llvm-project.git 78f1a82fa1e508c852d1ea65caf28ec43a1b74b4)
[2022-08-24T18:41:39.677Z] Target: x86_64-unknown-linux-gnu
[2022-08-24T18:41:39.677Z] Thread model: posix
[2022-08-24T18:41:39.677Z] InstalledDir: /llvm-project/build/bin
[2022-08-24T18:41:39.937Z] clang-13: note: diagnostic msg: Error generating preprocessed source(s).
```

**Comment by rosenrodt** [comment]
```
'library/src/tensor_operation_instance/gpu/batched_gemm_gemm/CMakeFiles/device_batched_gemm_gemm_instance.dir/device_batched_gemm_gemm_xdl_cshuffle_f16_f16_f16_f16_gmk_gnk_gno_gmo_instance.cpp.o' failed
```

Restarted CI and still no success. Looks like we break the compiler again

**Comment by asroy** [comment]
no longer needed

**Pull request closed without merging**